client/starwhale/api/_impl/model.py

```diff
@@ -316,86 +257,47 @@ def _starwhale_internal_run_ppl(self) -> None:
-                self._sw_logger.error(msg)
-                if not self.ignore_error:
-                    raise Exception(msg)
-
+        for _idx, _data, _annotations in _data_loader:
             pred: t.Any = b""
             exception = None
             try:
                 # TODO: inspect profiling
-                pred = self.ppl(
-                    data.data.encode() if isinstance(data.data, str) else data.data,
-                    data_index=data.idx,
-                    data_size=data.data_size,
-                    label_content=label.data,
-                    label_size=label.data_size,
-                    label_index=label.idx,
-                    ds_name=data.ext_attr.get("ds_name", ""),
-                    ds_version=data.ext_attr.get("ds_version", ""),
-                )
+                pred = self.ppl(_data, annotations=_annotations, index=_idx)
             except Exception as e:
                 exception = e
-                self._sw_logger.exception(f"[{data.idx}] data handle -> failed")
+                self._sw_logger.exception(f"[{_idx}] data handle -> failed")
                 if not self.ignore_error:
                     self._update_status(STATUS.FAILED)
                     raise
             else:
                 exception = None
 
-            self._do_record(data, label, exception, *pred)
+            self._do_record(_idx, _annotations, exception, *pred)
 
     def _do_record(
         self,
-        data: DataField,
-        label: DataField,
-        exception: t.Union[None, Exception],
+        idx: int,
+        annotations: t.Dict,
+        exception: t.Optional[Exception],
         *args: t.Any,
     ) -> None:
         _timeline = {
             "time": now_str(),
             "status": exception is None,
             "exception": str(exception),
-            "index": data.idx,
-            self._ppl_data_field: base64.b64encode(
-                self.ppl_data_serialize(*args)
-            ).decode("ascii"),
+            "index": idx,
         }
         self._timeline_writer.write(_timeline)
 
-        _label = ""
-        if self.merge_label:
-            try:
-                label = self.handle_label(
-                    label.data,
-                    index=label.idx,
-                    size=label.data_size,
-                )
-                _label = base64.b64encode(self.label_data_serialize(label)).decode(
-                    "ascii"
-                )
-            except Exception as e:
-                self._sw_logger.exception(f"{label.data!r} label handle exception:{e}")
-                if not self.ignore_error:
-                    self._update_status(STATUS.FAILED)
-                    raise
-                else:
-                    _label = ""
-        # self._sw_logger.debug(f"record ppl result:{data.idx}")
+        annotations = {} if self.ignore_annotations else annotations
+        _b64: t.Callable[[bytes], str] = lambda x: base64.b64encode(x).decode("ascii")
```

Comment:
is it better to move this to class or global scope?

Reply:
base64 encode is already a function. We use a lambda function only as a shortcut.

Reply:
ok